---
title: "scDist: Robust identification of perturbed cell types in single-cell RNA-seq data"
author: "R package version 1.1.0"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## System Requirements

`R` is required to use `scDist`. In development, `R` version 4.0.0 and greater were used, but there may be compatibility with previous versions. 

## Installation 

From the R console, `devtools::install_github("phillipnicol/scDist")`. Installation should take less than a minute on a standard machine. 

## Demo 

The input to `scDist` is a normalized count matrix and correpsonding metadata that describes what condition and patient each cell belongs to. In this demo, we create a simulated dataset with 10 cell types. The demo should take less than a minute to run on a standard machine. The code is also modifiable to see how `scDist` performs for different parameter values.

```{r,message=FALSE,warning=FALSE}
library(scDist)
set.seed(1126490984)
```

Generate simulated data with 10 cell types and 5 patients in each group: 

```{r}
sim <- simData(nct=10,N1=5,N2=5)

dim(sim$Y) #Normalized counts
rownames(sim$Y) <- 1:1000
head(sim$meta.data)
```
Now we apply scDist: 

```{r}
out <- scDist(sim$Y,sim$meta.data,fixed.effects = "response",
              random.effects="patient",
              clusters="clusters")
```

The results data frame gives a summary of the estimated distance and uncertainty for each cell type 

```{r}
out$results
```

The true distances are

```{r}
names(sim$D.true) <- letters[1:length(sim$D.true)]
sim$D.true
```


We can also plot the results

```{r}
DistPlot(out)
```

To get a plot of genes that are associated with the perturbation use `distGenes`:

```{r}
distGenes(out, cluster = "a")
```



## Reference

If you use `scDist` in your work, please cite:

Nicol, P.B., Paulson, D., Qian, G., Liu, X.S., Irizarry, R.A., and Sahu, A.D. (2023). Robust identification of perturbed cell types in single-cell RNA-seq data. bioRxiv.


